src/rsa/padding.rs

```diff
@@ -103,3 +103,119 @@ impl Verification for PKCS1 {
+        encoded.read_all((), |em| {
+            let digest_len = self.digest_alg.output_len;
+
+            // Step 2.
+            let m_hash = digest::digest(self.digest_alg,
+                                        msg.as_slice_less_safe());
+
+            // Step 3: where we assume the digest and salt are of equal length.
+            if em_len < 2 + (2 * digest_len) {
+                return Err(())
+            }
+
+            // Steps 4 and 5: Parse encoded message as:
+            //     masked_db || h_hash || 0xbc
+            let db_len = em_len - digest_len - 1;
+            let masked_db = try!(em.skip_and_get_input(db_len));
+            let h_hash = try!(em.skip_and_get_input(digest_len));
+            if try!(em.read_byte()) != 0xbc {
+                return Err(());
+            }
+
+            // Step 7.
+            let mut db = [0u8; MAX_OUTPUT_LEN];
+            try!(mgf1(h_hash.as_slice_less_safe(), db_len, &mut db,
+                      self.digest_alg));
+
+            try!(masked_db.read_all((), |masked_bytes| {
+                // Step 6. Moduli lengths are always a multiple of 8 so the top
+                // bit must be 0.
+                let b = try!(masked_bytes.read_byte());
+                if b & 0x80 != 0 {
+                    return Err(());
+                } else {
+                    db[0] ^= b;
+                }
+
+                // Step 8.
+                for i in 1..db_len {
+                    let b = try!(masked_bytes.read_byte());
+                    db[i] ^= b;
+                }
+                Ok(())
+            }));
+
+            // Step 9.
+            db[0] = db[0] & 0x7f;
+
+            // Step 10.
+            let pad_len = db_len - digest_len - 1;
+            for i in 0..pad_len {
+                if db[i] != 0 {
+                    return Err(());
+                }
+            }
+            if db[pad_len] != 1 {
+                return Err(());
+            }
+
+            // Step 11.
+            let salt = &db[db_len - digest_len..][..digest_len];
+
+            // Step 12 and 13: compute hash value of:
+            //     (0x)00 00 00 00 00 00 00 00 || m_hash || salt
+            let mut ctx = digest::Context::new(self.digest_alg);
+            let prefix = [0u8; 8];
+            ctx.update(&prefix);
+            ctx.update(m_hash.as_ref());
+            ctx.update(&salt);
+            let h_hash_check = ctx.finish();
+
+            // Step 14.
+            if h_hash != h_hash_check.as_ref() {
+                return Err(());
+            }
+
+            Ok(())
+        })
+    }
+}
+
+// Mask-generating function MGF1 as described in:
+// https://tools.ietf.org/html/rfc3447#appendix-B.2.1
+fn mgf1(seed: &[u8], mask_len: usize, mask: &mut [u8],
+        digest_alg: &'static digest::Algorithm) -> Result<(), ()> {
+
+    let digest_len = digest_alg.output_len;
+
+    let ctr_max = 1 + ((mask_len - 1) / digest_len);
+    for i in 0..ctr_max {
```

Comment:
Maybe consider just putting the expression in the for loop header, rather than using a temporary variable binding? Also, the value of `ctr_max` looks like it will use truncating division at the face of it, maybe add a comment explaining the preconditions for these numbers?